### PR TITLE
feat: password authentication

### LINKED ISSUE #81: Is it possible to connect with password to redis db?
Is it authorization supported?

### PR DESCRIPTION
**Pull request by uki00a**
This PR resolves #81.
Please review @keroxp 🙏, cc @devalexqt.

- Added the `password` property to `RedisConnectOptions`.
- `AUTH` command is now executed automatically when password is specified.

### CODE REVIEW

**Comment by devalexqt** [comment]
How I can test it?

**Comment by uki00a** [comment]
@devalexqt You can import the source modified by this PR from here: https://deno.land/x/redis@1025f44f6a5fa8d5510fc7a56d27abe67f868d6b/mod.ts

```typescript
import { connect } from "https://deno.land/x/redis@1025f44f6a5fa8d5510fc7a56d27abe67f868d6b/mod.ts";

const redis = await connect(
  {
    hostname: "localhost",
    port: 6379,
    password: "foobar"
  },
);
doSomethingWithRedis(redis);
```

**Comment by devalexqt** [comment]
Thanks now it's working with password.

File: redis.ts [hunk 1]
```diff
@@ -1504,6 +1506,14 @@ export async function connect({
   const bufw = new BufWriter(conn);
   const exec = muxExecutor(bufr, bufw);
   const client = await create(conn, conn, conn, exec);
+  if (password) {
```
Comment (keroxp) [inline redis.ts:1509]: Is empty string acceptable? I don't know redis support empty password for authentication, I think we should use `!= null` operator.

File: redis.ts [hunk 2]
```diff
@@ -1504,6 +1506,14 @@ export async function connect({
...
+    try {
+      await client.auth(password);
+    } catch (err) {
+      client.close();
+      return Promise.reject(err);
```
Comment (keroxp) [inline redis.ts:1514]: `throw` is natural IMO

File: tests/general_test.ts [hunk 3]
```diff
@@ -35,6 +37,15 @@ test("db0", async function testDb0Option() {
   client2.close();
 });
 
+test("connect with wrong password", async function testConnectWithPassword() {
```
Comment (keroxp) [inline tests/general_test.ts:40]: Could you add test case for empty string password?

File: redis.ts (hunk 1 above)
Comment (uki00a) [reply redis.ts:1509]: Yes, it is the following reasons:

- In JavaScript, an empty string is evaluated as `false`.
- In Redis, setting `requirepass` to an empty string disables password authentication.

```shell
$ redis-cli
127.0.0.1:6379> config get requirepass
1) "requirepass"
2) ""
127.0.0.1:6379> config set requirepass foobar
OK
127.0.0.1:6379> quit


$ redis-cli
127.0.0.1:6379> exists "hoge"
(error) NOAUTH Authentication required.
127.0.0.1:6379> auth foobar
OK
127.0.0.1:6379> exists "hoge"
(integer) 0
127.0.0.1:6379> config set requirepass ""
OK
127.0.0.1:6379> quit


$ redis-cli
127.0.0.1:6379> exists "hoge"
(integer) 0
```

Comment (uki00a) [reply redis.ts]: In Redis, empty string is always denied.

So, Therefore, your point is correct :+1:

**Comment by uki00a** [comment]
Thanks!
